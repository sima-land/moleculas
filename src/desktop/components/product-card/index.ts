export { ProductCardProps } from './types';
export { ProductCard, HoverCard, reduceBaseInfo, reduceHoverInfo } from './product-card';
