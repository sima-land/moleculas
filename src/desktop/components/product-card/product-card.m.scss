.root {
  position: relative;
  max-width: 400px;
}

.card {
  position: absolute;
  top: -16px;
  left: -16px;
  padding: 16px 16px 24px;
  width: calc(100% + 32px);
}
