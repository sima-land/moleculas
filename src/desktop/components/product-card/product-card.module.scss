@use 'node_modules/@sima-land/ui-nucleons/colors';

.root {
  padding: 16px 16px 24px;
  background: #fff;
}

.quick-view-button {
  position: absolute;
  top: 8px;
  right: 8px;
  cursor: pointer;
}

.control-wrapper {
  display: flex;
  flex-wrap: wrap;
  margin-top: 16px;
  &.loading {
    .control {
      background: colors.$gray4;
      border-radius: 4px;
    }
    .unit-text::before {
      content: '';
      display: block;
      background: colors.$gray4;
      border-radius: 4px;
      width: 48px;
      height: 16px;
    }
  }
}

.control {
  width: 122px;
  height: 40px;
  margin-right: 16px;
}

.unit-text {
  padding-top: 12px; // чтобы при переносе на следующую строку правильно отступил от кнопки
  font-size: 12px;
  line-height: 16px;
  color: colors.$gray38;
}

.bottom-text {
  margin-top: 12px;
  font-size: 12px;
  line-height: 16px;
  color: colors.$gray38;
}
