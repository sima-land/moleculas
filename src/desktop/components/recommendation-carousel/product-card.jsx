import React, { forwardRef } from 'react';
import { BoxShadow as Shadow } from '@dev-dep/ui-nucleons/styling/shadows';
import { MediumRounds as Rounds } from '@dev-dep/ui-nucleons/styling/shapes';
import { ProductInfo } from './product-info';
import { InCartControl } from './in-cart-control';
import classnames from 'classnames/bind';
import styles from './product-card.scss';

const cx = classnames.bind(styles);

/**
 * Карточка рекомендованного товара.
 */
export const ProductCard = forwardRef(({
  className,
  inCartControlProps,
  productInfo,
  ...divProps
}, ref) => (
  <div
    {...divProps}
    ref={ref}
    className={cx('root', Rounds.all, Shadow.z4, className)}
  >
    <ProductInfo
      {...productInfo}
      withImageButtons
    />

    <InCartControl
      {...inCartControlProps}
      className={cx('stepper')}
    />

    {productInfo && productInfo.markupText && (
      <div className={cx('bottom-text')}>
        {productInfo.markupText}
      </div>
    )}
  </div>
));
