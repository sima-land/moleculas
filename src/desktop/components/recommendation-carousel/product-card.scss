@use 'node_modules/@dev-dep/ui-nucleons/colors';

.root {
  padding: 16px 16px 24px;
  background: #fff;
}

.stepper {
  margin-top: 16px;
}

.bottom-text {
  margin-top: 12px;
  font-size: 12px;
  line-height: 16px;
  color: colors.$gray38;
}
