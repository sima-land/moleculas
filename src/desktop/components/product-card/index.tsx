import React, { Children, forwardRef, isValidElement, cloneElement, useState } from 'react';
import { ProductInfo, ProductInfoProps, Parts } from '../../../common/components/product-info';
import { Plate, PlateProps } from '@sima-land/ui-nucleons/plate';
import { useLayer } from '@sima-land/ui-nucleons/helpers/layer';
import cn from 'classnames';
import styles from './product-card.module.scss';

export type ProductCardChildren = React.ReactElement<ProductInfoProps, typeof ProductInfo>;

export interface ProductCardProps extends React.HTMLAttributes<HTMLDivElement> {
  children: ProductCardChildren;
}

export const HoverCard = forwardRef<HTMLDivElement | null, PlateProps>((props, ref) => {
  const layer = useLayer() + 1;

  return (
    <Plate
      {...props}
      style={{ ...props.style, zIndex: layer }}
      ref={ref}
      rounds='m'
      shadow='z4'
      className={cn(styles.card, props.className)}
      data-testid='product-card:hover-card'
    />
  );
});

export const ProductCard = ({ children, className, ...props }: ProductCardProps) => {
  const [hovered, toggle] = useState<boolean>(false);

  return (
    <div {...props} data-testid='product-card' className={cn(styles.root, className)}>
      <div data-testid='product-card:info' onMouseEnter={() => toggle(true)}>
        {reduceBaseInfo(children)}
      </div>

      {hovered && (
        <HoverCard onMouseLeave={() => toggle(false)}>{reduceHoverInfo(children)}</HoverCard>
      )}
    </div>
  );
};

/**
 * Фильтрует содержимое карточки для вывода состояния без наведения.
 * @param el Элемент.
 * @return Элемент.
 */
export const reduceBaseInfo = (el: ProductCardChildren) =>
  cloneElement(el, {
    children: Children.map(el.props.children, child => {
      let result;

      if (isValidElement(child)) {
        switch (child.type) {
          // игнорируем кнопки-иконки на изображении
          case Parts.Image:
            result = cloneElement<any>(child, { children: undefined });
            break;

          // игнорируем футер
          case Parts.Footer:
            result = null;
            break;

          // остальное оставляем без изменений
          default:
            result = child;
            break;
        }
      }

      return result;
    }),
  });

export const reduceHoverInfo = (el: ProductCardChildren) =>
  cloneElement(el, {
    // при наведении у изображения строго непрозрачность = 1
    children: Children.map(el.props.children, child =>
      isValidElement(child) && child.type === Parts.Image
        ? cloneElement(child, { opacity: 1 })
        : child,
    ),
  });
