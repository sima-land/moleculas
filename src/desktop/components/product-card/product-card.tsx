import { Children, forwardRef, isValidElement, cloneElement, useState } from 'react';
import { ProductCardChildren, ProductCardProps, ReduceBaseInfoOptions } from './types';
import { Parts } from '../../../common/components/product-info';
import { Plate, PlateProps } from '@sima-land/ui-nucleons/plate';
import { useLayer } from '@sima-land/ui-nucleons/helpers';
import classNames from 'classnames/bind';
import styles from './product-card.module.scss';

const cx = classNames.bind(styles);

/**
 * Карточка данных товара с расширенной инф-ой при наведении.
 * @param props Свойства.
 * @return Элемент.
 */
export function ProductCard({
  // @todo не совсем правильно заставлять работать с внутренностями реакта - надо придумать что-то более простое
  reduceBaseInfo: reduceBaseInfoProp = reduceBaseInfo,
  reduceHoverInfo: reduceHoverInfoProp = reduceHoverInfo,
  children,
  className,
  ...props
}: ProductCardProps) {
  const [hovered, toggle] = useState<boolean>(false);

  return (
    <div
      {...props}
      data-testid='product-card'
      className={cx('root', className)}
      onMouseEnter={() => toggle(true)}
      // ВАЖНО: вешаем onMouseLeave на корневой элемент так как в Chrome проблемы если вешать на "product-card:info"
      onMouseLeave={() => toggle(false)}
    >
      <div data-testid='product-card:info'>{reduceBaseInfoProp(children)}</div>

      {hovered && <HoverCard>{reduceHoverInfoProp(children)}</HoverCard>}
    </div>
  );
}

export const HoverCard = forwardRef<HTMLDivElement | null, PlateProps>((props, ref) => {
  const layer = useLayer() + 1;

  return (
    <Plate
      {...props}
      style={{ ...props.style, zIndex: layer }}
      ref={ref}
      rounds='m'
      shadow='z4'
      className={cx('card', props.className)}
      data-testid='product-card:hover-card'
    />
  );
});

/**
 * Фильтрует содержимое карточки для вывода состояния без наведения.
 * @param element Элемент.
 * @param options Опции.
 * @return Элемент.
 */
export function reduceBaseInfo(
  element: ProductCardChildren,
  { hideImageButtons = true, hideFooter = true }: ReduceBaseInfoOptions = {},
) {
  return cloneElement(element, {
    children: Children.map(element.props.children, child => {
      let result;

      if (isValidElement(child)) {
        switch (true) {
          // игнорируем или фильтруем кнопки-иконки на изображении
          case Boolean(child.type === Parts.Image && hideImageButtons):
            result = cloneElement<any>(child, {
              children:
                typeof hideImageButtons === 'function'
                  ? Children.toArray(child.props.children).filter(hideImageButtons)
                  : undefined,
            });
            break;

          // игнорируем футер
          case Boolean(child.type === Parts.Footer && hideFooter):
            result = null;
            break;

          // остальное оставляем без изменений
          default:
            result = child;
            break;
        }
      }

      return result;
    }),
  });
}

/**
 * Получив содержимое карточки фильтрует/меняет его для вывода в карточке при наведении.
 * @param element Содержимое.
 * @return Элемент.
 */
export function reduceHoverInfo(element: ProductCardChildren) {
  return cloneElement(element, {
    children: Children.map(element.props.children, child => {
      let result;

      if (isValidElement(child)) {
        switch (child.type) {
          // при наведении у изображения строго непрозрачность = 1
          case Parts.Image:
            result = cloneElement<any>(child, { opacity: 1 });
            break;

          // остальное оставляем без изменений
          default:
            result = child;
            break;
        }
      }

      return result;
    }),
  });
}
