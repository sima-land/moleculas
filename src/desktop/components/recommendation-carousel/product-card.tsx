import React, { forwardRef } from 'react';
import { BoxShadow as Shadow } from '@sima-land/ui-nucleons/styling/shadows';
import { MediumRounds as Rounds } from '@sima-land/ui-nucleons/styling/shapes';
import { ProductInfo, ProductInfoProps } from './product-info';
import { InCartControl, InCartControlProps } from '../../../common/components/in-cart-control';
import classnames from 'classnames/bind';
import styles from './product-card.module.scss';
import { CartItem } from './types';

export interface ProductCardProps extends React.HTMLAttributes<HTMLDivElement> {
  inCartControl: CartItem & InCartControlProps
  productInfo: Omit<ProductInfoProps, 'withQuickView'>
}

const cx = classnames.bind(styles);

/**
 * Карточка рекомендованного товара.
 */
export const ProductCard = forwardRef<HTMLDivElement | null, ProductCardProps>(({
  className,
  inCartControl,
  productInfo,
  ...divProps
}, ref) => (
  <div
    {...divProps}
    ref={ref}
    className={cx('root', Rounds.all, Shadow.z4, className)}
  >
    <ProductInfo
      {...productInfo}
      withQuickView
    />

    <div className={cx('control-wrapper')}>
      <InCartControl {...inCartControl} className={cx('control')} />

      {inCartControl && inCartControl.stepText && (
        <div className={cx('unit-text')}>
          {inCartControl.stepText}
        </div>
      )}

      {inCartControl && inCartControl.markupText && (
        <div className={cx('bottom-text')}>
          {inCartControl.markupText}
        </div>
      )}
    </div>
  </div>
));
