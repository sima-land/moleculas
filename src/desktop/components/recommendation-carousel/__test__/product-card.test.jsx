import React from 'react';
import { mount } from 'enzyme';
import { ProductCard } from '../product-card';

describe('<ProductCard />', () => {
  it('should renders correctly', () => {
    const wrapper = mount(
      <ProductCard
        productInfo={{
          url: 'https://www.sima-land.ru/123456',
          name: 'Ножницы портновские, 10, 26 см, цвет чёрный',
          imageSrc: 'https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v: 1617955476',
          price: 225,
          oldPrice: 250,
          favorite: false,
          currencyGrapheme: '₽',
          withImageButtons: true,
          onQuickViewClick: jest.fn(),
          onFavoriteClick: jest.fn(),
          markupText: 'Комплектация + 50 ₽ при покупке до 20 шт',
        }}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
