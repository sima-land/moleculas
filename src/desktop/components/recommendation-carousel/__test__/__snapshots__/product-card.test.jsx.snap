// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> should renders correctly 1`] = `
<ForwardRef
  productInfo={
    Object {
      "currencyGrapheme": "₽",
      "favorite": false,
      "imageSrc": "https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v: 1617955476",
      "markupText": "Комплектация + 50 ₽ при покупке до 20 шт",
      "name": "Ножницы портновские, 10, 26 см, цвет чёрный",
      "oldPrice": 250,
      "onFavoriteClick": [MockFunction],
      "onQuickViewClick": [MockFunction],
      "price": 225,
      "url": "https://www.sima-land.ru/123456",
      "withImageButtons": true,
    }
  }
>
  <div
    className="root rounds-md__all shadow-z4"
  >
    <ProductInfo
      currencyGrapheme="₽"
      favorite={false}
      imageSrc="https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v: 1617955476"
      markupText="Комплектация + 50 ₽ при покупке до 20 шт"
      name="Ножницы портновские, 10, 26 см, цвет чёрный"
      oldPrice={250}
      onFavoriteClick={[MockFunction]}
      onQuickViewClick={[MockFunction]}
      price={225}
      url="https://www.sima-land.ru/123456"
      withImageButtons={true}
    >
      <a
        className="row image-row"
        href="https://www.sima-land.ru/123456"
      >
        <Image
          alt="Ножницы портновские, 10, 26 см, цвет чёрный"
          className="image"
          src="https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v: 1617955476"
        >
          <div
            className="root image"
          >
            <img
              alt="Ножницы портновские, 10, 26 см, цвет чёрный"
              src="https://cdn3.static1-sima-land.com/items/29455/0/1600.jpg?v: 1617955476"
            />
          </div>
        </Image>
        <WishButton
          className="on-image-btn wish"
          isWished={false}
          onClick={[Function]}
        >
          <div
            className="wish-button on-image-btn wish"
            data-testid="favorite-button"
            onClick={[Function]}
            title="Добавить в избранное"
          >
            <HeartSVG
              className="svg stroke"
              height={24}
              viewBox="0 0 24 24"
              width={24}
              xmlns="http://www.w3.org/2000/svg"
            >
              <svg
                className="svg stroke"
                height={24}
                viewBox="0 0 24 24"
                width={24}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19.944 3.6h-.001c-2.7-1.879-5.978-1.165-7.943.793-1.964-1.958-5.241-2.683-7.943-.792C2.524 4.673 1.57 6.473 1.504 8.369c-.075 2.145.826 4.042 2.36 5.985 1.512 1.912 3.685 3.927 6.261 6.315l.086.08h.002l.098.091a2.466 2.466 0 0 0 3.371-.012l.107-.1.027-.025c2.602-2.407 4.795-4.436 6.316-6.361 1.534-1.941 2.435-3.838 2.364-5.982v-.001c-.065-1.886-1.02-3.686-2.552-4.758zM12 18.274c-2.278-2.105-4.14-3.84-5.45-5.401-1.358-1.62-2.047-2.968-2.047-4.261 0-1.774 1.285-3.07 2.999-3.07 1.055 0 2.103.553 2.72 1.39.38.516.996 1.017 1.78 1.017.789 0 1.406-.507 1.787-1.025.612-.831 1.657-1.382 2.71-1.382 1.713 0 2.998 1.296 2.998 3.07 0 1.293-.689 2.641-2.047 4.26-1.309 1.561-3.171 3.296-5.45 5.402z"
                />
              </svg>
            </HeartSVG>
          </div>
        </WishButton>
        <QuickViewButton
          className="on-image-btn quick-view"
          onClick={[Function]}
        >
          <div
            className="button on-image-btn quick-view"
            data-testid="quick-view-button"
            onClick={[Function]}
            title="Быстрый просмотр"
          >
            <QuickViewSVG
              className="stroke"
              height={24}
              viewBox="0 0 24 24"
              width={24}
              xmlns="http://www.w3.org/2000/svg"
            >
              <svg
                className="stroke"
                height={24}
                viewBox="0 0 24 24"
                width={24}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19.94 22.06l-3.219-3.217-.339.22A8.956 8.956 0 0 1 11.5 20.5a9 9 0 1 1 9-9c0 1.8-.528 3.476-1.438 4.882l-.22.34 3.219 3.217a1.5 1.5 0 0 1-2.122 2.122zM9 10h1V9a1.5 1.5 0 1 1 3 0v1h1a1.5 1.5 0 1 1 0 3h-1v1a1.5 1.5 0 1 1-3 0v-1H9a1.5 1.5 0 1 1 0-3zm2.5-4.5a6 6 0 1 0 0 12 6 6 0 0 0 0-12z"
                />
              </svg>
            </QuickViewSVG>
          </div>
        </QuickViewButton>
      </a>
      <div
        className="row prices-row"
      >
        <Price
          className="price"
          currencyGrapheme="₽"
          value={225}
        >
          <span
            className="root price"
          >
            225 ₽
          </span>
        </Price>
        <Price
          className="old-price"
          crossedOut={true}
          currencyGrapheme="₽"
          value={250}
        >
          <span
            className="root old-price crossed-out"
          >
            250 ₽
          </span>
        </Price>
      </div>
      <ForwardRef(Link)
        className="row link"
        color="gray87"
        href="https://www.sima-land.ru/123456"
      >
        <a
          className="link row link color__gray87 hover-color__gray54"
          href="https://www.sima-land.ru/123456"
        >
          Ножницы портновские, 10, 26 см, цвет чёрный
        </a>
      </ForwardRef(Link)>
    </ProductInfo>
    <InCartControl
      className="stepper"
    >
      <div
        className="root stepper"
      >
        <ForwardRef(Button)
          className="button"
          size="small"
        >
          <button
            className="button button-base button-primary button-small text-button-small"
          >
            В корзину
          </button>
        </ForwardRef(Button)>
        <div
          className="unit-text"
        >
          По 1 шт
        </div>
      </div>
    </InCartControl>
    <div
      className="bottom-text"
    >
      Комплектация + 50 ₽ при покупке до 20 шт
    </div>
  </div>
</ForwardRef>
`;
